src/sections/image_resources_section.rs

```diff
@@ -111,41 +214,522 @@ impl ImageResourcesSection {
+        let name = cursor.read_unicode_string()?;
+        let class_id = DescriptorStructure::read_key_length(cursor)?.to_vec();
+
+        Ok(ClassStructure {
+            name,
+            class_id,
+        })
+    }
+
+    fn read_enumerated_reference(cursor: &mut PsdCursor) -> Result<EnumeratedReference, Error> {
+        let name = cursor.read_unicode_string()?;
+        let class_id = DescriptorStructure::read_key_length(cursor)?.to_vec();
+        let key_id = DescriptorStructure::read_key_length(cursor)?.to_vec();
+        let enum_field = DescriptorStructure::read_key_length(cursor)?.to_vec();
+
+        Ok(EnumeratedReference {
+            name,
+            class_id,
+            key_id,
+            enum_field,
+        })
+    }
+
+    fn read_offset_structure(cursor: &mut PsdCursor) -> Result<OffsetStructure, Error> {
+        let name = cursor.read_unicode_string()?;
+        let class_id = DescriptorStructure::read_key_length(cursor)?.to_vec();
+        let offset = cursor.read_u32()?;
+
+        Ok(OffsetStructure {
+            name,
+            class_id,
+            offset,
+        })
     }
 
-    fn read_descriptor_structure(cursor: &mut PsdCursor) -> Result<(), Error> {
-        // TODO: make descriptor decoder
-        Ok(())
+    fn read_alias_structure(cursor: &mut PsdCursor) -> Result<AliasStructure, Error> {
+        let length = cursor.read_u32()?;
+        let data = cursor.read(length)?.to_vec();
+
+        Ok(AliasStructure {
+            data
+        })
+    }
+
+    fn read_list_structure(cursor: &mut PsdCursor) -> Result<Vec<DescriptorField>, Error> {
+        DescriptorStructure::read_list(cursor, true)
+    }
+
+    fn read_enumerated_descriptor(cursor: &mut PsdCursor) -> Result<EnumeratedDescriptor, Error> {
+        let type_field = DescriptorStructure::read_key_length(cursor)?.to_vec();
+        let enum_field = DescriptorStructure::read_key_length(cursor)?.to_vec();
+
+        Ok(EnumeratedDescriptor {
+            type_field,
+            enum_field,
+        })
+    }
+
+    fn read_raw_data(cursor: &mut PsdCursor) -> Result<Vec<u8>, Error> {
+        let length = cursor.read_u32()?;
+        Ok(cursor.read(length)?.to_vec())
+    }
+
+    // Note: this structure is not documented
+    fn read_name(cursor: &mut PsdCursor) -> Result<NameStructure, Error> {
+        let name = cursor.read_unicode_string()?;
+        let class_id = DescriptorStructure::read_key_length(cursor)?.to_vec();
+        let value = cursor.read_unicode_string()?;
+
+        Ok(NameStructure {
+            name,
+            class_id,
+            value,
+        })
+    }
+
+    fn read_key_length<'a>(cursor: &'a mut PsdCursor) -> Result<&'a [u8], Error> {
+        let length = cursor.read_u32()?;
+        let length = if length > 0 {
+            length
+        } else {
+            4
+        };
+
+        let key = cursor.read(length)?;
+        Ok(key)
     }
 }
+
+#[cfg(test)]
```

Comment:
It's hard to tell what these descriptors are and what we're testing here. 

It appears as if we're just making sure that they don't panic.

Can we:

1. Rename the files to describe what type of data that they hold

2. `let some_var_name = DescriptorStructure::read_descriptor_structure(...).unwrap(); assert_eq!(some_var_name.some_val, expected);`